subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/AbstractFileCollectionSnapshotter.java

```diff
@@ -91,15 +90,14 @@ public void visitCollection(FileCollectionInternal fileCollection) {
 
         @Override
         public void visitTree(FileTreeInternal fileTree) {
-            List<FileSnapshot> descendants = fileSystemSnapshotter.snapshotTree(fileTree);
-            fileSnapshotVisitor.visitFileTreeSnapshot(descendants);
+            VisitableDirectoryTree treeSnapshot = fileSystemSnapshotter.snapshotTree(fileTree);
+            fileSnapshotVisitor.visitFileTreeSnapshot(treeSnapshot);
```

Comment:
Would it make sense to turn this around into `treeSnapshot.visit(...)`?

Also, `treeSnapshot` is not really a snapshot of the tree anymore, it's just the entry point you can use to traverse from. I think we should rename it accordingly.